### PR TITLE
BUG: incorrect set_labels in MI

### LINKED ISSUE #19057: Bug in set_labels of MultiIndex
```python
import pandas as pd
                                 
idx = pd.MultiIndex.from_tuples([(0, i) for i in range(130)])
print idx.labels[1].dtype  
# --> int16
idx.set_labels(labels=idx.labels[1], level=1, inplace=True) 
print idx.labels[1].dtype  
# --> int8
```

Imho this is caused by the following line, because the erratic assumption that the indices of `labels` and `self.levels` would be index-aligned.
https://github.com/pandas-dev/pandas/blame/master/pandas/core/indexes/multi.py#L331
Hence the call to `coerce_indexer_dtype` casts the given `labels` to int8. This is because its l'ooking at the number of categories in level 0.

I would suggest to replace 
```python
            for l, lev, lab in zip(level, self.levels, labels):
                new_labels[l] = _ensure_frozen(
                    lab, lev, copy=copy)._shallow_copy()
```
with
```python
            for l, lab in zip(level, labels):
                lev = self.levels[l]
                new_labels[l] = _ensure_frozen(
                    lab, lev, copy=copy)._shallow_copy()
```

### PR DESCRIPTION
**Pull request by Mofef**
This fixes #19057

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
this looks ok, but pls add your original test (both with and w/o inplace) as well as a note in the whatsnew

**Comment by jreback** [comment]
cc @toobaz 

any thoughts here

**Comment by toobaz** [comment]
Looks good to me. As for tests, I guess it's just a matter of understanding what went wrong in [this one](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/indexes/test_multi.py#L213), and fix it.

**Comment by Mofef** [comment]
you probably mean https://github.com/pandas-dev/pandas/blob/93033151a8d8aaa650a81df9f41347758bf6c393/pandas/tests/indexes/test_multi.py#L287 , set_levels works correctly iirc.
The issue with the test is that self.index has the same magnitude of categories in each level. (i.e. the length is each smaller than 127).  So the missalignment dos not have an effect. Also there is no missalignment if the labels get replaced in the first n levels. 
I would like to take care of the tests later.

**Comment by toobaz** [comment]
> you probably mean

yep, sorry, actually this assertion:
https://github.com/pandas-dev/pandas/blob/93033151a8d8aaa650a81df9f41347758bf6c393/pandas/tests/indexes/test_multi.py#L302

But I agree with your explanation.

**Comment by jreback** [comment]
@Mofef tests need to be in the same issue as the PR fixing.

**Comment by Mofef** [comment]
@jreback Sorry I don't understand, you mean i should also add tests to this PR?
In the meanwhile i looked into test_multi.py. My initial idea was to just increase the size of the minor level with something like 
```python
    def setup_method(self, method):
        major_axis = Index(['foo', 'bar', 'baz', 'qux'])
        minor_axis = Index(range(2, 128))

        major_labels = np.random.choice(range(4), 128)
        minor_labels = np.array(range(128))
...
```
However, that doesn't seem to be feasible since a lot of test methods rely on the current "reference index". So I get ~50 presumably false negative tests. 
Another option would be to create a bigger `MultiIndex` down in `test_set_labels`. On the other hand it's probably a good idea to check each operation against such an index with `levels` of different `dtype`.
What do you think?

**Comment by jreback** [comment]
@Mofef yes you need a test, you can use the one that you put in the issue. you need to assert that this works (eg. construct the expected MI).

**Comment by Mofef** [comment]
Done. I had to define another test method, with its own `assert_matching` function, since inside `test_set_labels` the expected labels get casted to int8. Imho this should happen outside of a function called "assert_equal"

I also stumbled upon #19092

File: pandas/tests/indexes/test_multi.py [hunk 1]
```diff
@@ -327,6 +327,28 @@ def assert_matching(actual, expected):
...
+        def assert_equal(actual, expected):
+            assert len(actual) == len(expected)
+            for act, exp in zip(actual, expected):
+                act = np.asarray(act)
+                exp = np.asarray(exp)
+                tm.assert_numpy_array_equal(act, exp, check_dtype=True)
+
+        ind = pd.MultiIndex.from_tuples([(0, i) for i in range(130)])
+        new_labels = range(129, -1, -1)
+        ind_reference = pd.MultiIndex.from_tuples(
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: call this expected

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -327,6 +327,28 @@ def assert_matching(actual, expected):
...
+            [(0, i) for i in new_labels])
+
+        # [w/o  mutation]
+        ind2 = ind.set_labels(labels=new_labels, level=1)
+        assert_equal(ind2, ind_reference)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: call these result, then you can simply do

``assert result.equals(expected)``

for the inplace case you need to copy before

File: pandas/tests/indexes/test_multi.py [hunk 3]
```diff
@@ -327,6 +327,28 @@ def assert_matching(actual, expected):
         assert_matching(ind2.labels, new_labels)
         assert_matching(self.index.labels, labels)
 
+    def test_set_label_distinctly_sized_levels(self):
+        # label changing for levels of different magnitude of categories
+        def assert_equal(actual, expected):
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: see below this is not necessary

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Mofef** [comment]
Cool, thanks. PTAL

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good. ping on green.

**Comment by Mofef** [comment]
@jreback https://travis-ci.org/pandas-dev/pandas/jobs/325457802#L448
```
648.46s$ git fetch origin +refs/pull/19058/merge:
fatal: unable to access 'https://github.com/pandas-dev/pandas.git/': Failed to connect to github.com port 443: Connection timed out
```

**Comment by TomAugspurger** [comment]
Sorry about the extra commits @Mofef. I fixed a merge conflict in the release notes and the CI services apparently had an issue with that. They're running again now.

**Comment by Mofef** [comment]
Oh ok no problem. thank you :)

**Comment by jreback** [comment]
thanks @Mofef

**Pull request merged**